src/VisualStudio/Core/Def/Implementation/ProjectSystem/RunningDocumentTableEventTracker.cs

```diff
@@ -0,0 +1,192 @@
+                {
+                    _listener.OnRenameDocument(docCookie, pszMkDocumentNew, pszMkDocumentOld);
+                }
+            }
+
+            if ((grfAttribs & (uint)__VSRDTATTRIB3.RDTA_DocumentInitialized) != 0)
+            {
+                if (TryGetBuffer(docCookie, out var buffer))
+                {
+                    _listener.OnInitializedDocument(docCookie, _runningDocumentTable.GetDocumentMoniker(docCookie), buffer);
+                }
+            }
+
+            // When starting a diff, the RDT doesn't call OnBeforeDocumentWindowShow, but it does call
+            // OnAfterAttributeChangeEx for the temporary buffer. The native IDE used this even to
+            // add misc files, so we'll do the same.
+            if ((grfAttribs & (uint)__VSRDTATTRIB.RDTA_DocDataReloaded) != 0)
+            {
+                if (CheckPreconditions(docCookie))
+                {
+                    _listener.OnReloadDocumentData(docCookie, _runningDocumentTable.GetDocumentMoniker(docCookie));
+                }
+            }
+
+            if ((grfAttribs & (uint)__VSRDTATTRIB.RDTA_Hierarchy) != 0)
+            {
+                _listener.OnRefreshDocumentContext(docCookie, _runningDocumentTable.GetDocumentMoniker(docCookie));
+            }
+
+            return VSConstants.S_OK;
+        }
+
+        public int OnBeforeDocumentWindowShow(uint docCookie, int fFirstShow, IVsWindowFrame pFrame)
+        {
+            if (fFirstShow != 0)
+            {
+                if (TryGetBuffer(docCookie, out var buffer))
+                {
+                    _listener.OnBeforeOpenDocument(docCookie, _runningDocumentTable.GetDocumentMoniker(docCookie), buffer);
+                }
+            }
+
+            return VSConstants.S_OK;
+        }
+
+        public int OnAfterDocumentWindowHide(uint docCookie, IVsWindowFrame pFrame)
+        {
+            return VSConstants.E_NOTIMPL;
+        }
+
+        public int OnBeforeSave(uint docCookie)
+        {
+            return VSConstants.E_NOTIMPL;
+        }
+
+        /// <summary>
+        /// Gets the text buffer for a document cookie.
+        /// Also checks to make sure the document is initialized before returning.
+        /// </summary>
+        public bool TryGetBuffer(uint docCookie, out ITextBuffer textBuffer)
+        {
+            textBuffer = null;
+            if (!CheckPreconditions(docCookie))
+            {
+                return false;
+            }
+
+            if ((object)_runningDocumentTable.GetDocumentData(docCookie) is IVsTextBuffer bufferAdapter)
+            {
+                textBuffer = _editorAdaptersFactoryService.GetDocumentBuffer(bufferAdapter);
+                return true;
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Checks that we're on the UI thread and that the document has already been initialized.
+        /// </summary>
+        private bool CheckPreconditions(uint docCookie)
```

Comment:
Not sure if this is really helping code clarity: reading the code above made me think "what preconditions?" -- this isn't a wrapper for anything complicated...